### PR TITLE
GamepadTest: resize window, add gamepad list and labels

### PR DESCRIPTION
**Pull request by Geokureli**
- show input labels specific to the controller being used (e.g.: ABXY vs triangle square circle and X)
- show digital directions on analog sticks
- allow user to select radio buttons to switch the active gamepad being displayed
- bigger window size to show new info
- added misc graphical features

### CODE REVIEW

File: Input/GamepadTest/source/Main.hx [hunk 1]
```diff
@@ -8,6 +8,7 @@ class Main extends Sprite
 	public function new()
 	{
 		super();
-		addChild(new FlxGame(650, 440, PlayState));
+		addChild(new FlxGame(0, 0, PlayState));
+		trace("hello world");
```
Comment (Gama11) [inline Input/GamepadTest/source/Main.hx:12]: Leftover?

File: Input/GamepadTest/source/GamepadList.hx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	
+	function updateStyle()
+	{
+		for (button in getRadios())
+		{
+			button.button.up_color = FlxColor.BLACK;
+			button.button.over_color = FlxColor.BLACK;
+			button.button.down_color = FlxColor.BLACK;
+			button.button.getLabel().color = FlxColor.BLACK;
+			// button.textY = -4;
```
Comment (Gama11) [inline Input/GamepadTest/source/GamepadList.hx:84]: Leftover?

File: Input/GamepadTest/source/PlayState.hx [hunk 3]
```diff
@@ -23,27 +24,45 @@ class PlayState extends FlxState
...
 		createDisconnectedOverlay();
-		add(connectedGamepads = new FlxUIRadioGroup(500, 10, null, null, null, 25, 125, 25, 125));
+		
+		add(connectedGamepads = new GamepadList(FlxG.width - 150, 10, onGamepadChange));
+	}
+	
+	function createLabelToggle():Void
+	{
+		controller.inputLabels.visible = false;
+		var btn:FlxButton = null;
```
Comment (Gama11) [inline Input/GamepadTest/source/PlayState.hx:57]: `button`

File: Input/GamepadTest/source/PlayState.hx [hunk 4]
```diff
@@ -23,27 +24,45 @@ class PlayState extends FlxState
 	var attachmentLabel:FlxText;
 	var deadZoneStepper:FlxUINumericStepper;
 	var deadZoneModeDropDown:FlxUIDropDownMenu;
-	var connectedGamepads:FlxUIRadioGroup;
+	var connectedGamepads:GamepadList;
 	var disconnectedOverlay:FlxTypedGroup<FlxSprite>;
 	var gamepads:Array<FlxGamepad> = [];
+	var controller:Gamepad;
```
Comment (Gama11) [inline Input/GamepadTest/source/PlayState.hx:30]: `gamepad:Gamepad` makes more sense, no?

File: Input/GamepadTest/source/PlayState.hx [hunk 5]
```diff
@@ -118,31 +136,30 @@ class PlayState extends FlxState
...
 	{
-		super.update(elapsed);
-		updateConnectedGamepads();
-
-		var gamepad = FlxG.gamepads.lastActive;
-		if (gamepad == null)
-		{
+		if (controller.gamepad == null && gamepad != null)
+			setEnabled(true);
+		else if(FlxG.gamepads.numActiveGamepads == 0)
```
Comment (Gama11) [inline Input/GamepadTest/source/PlayState.hx:144]: Looks like you need to run haxe-formatter over this.

File: Input/GamepadTest/source/GamepadList.hx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+		super(x, y, ids, labels, null, 25, 200, 25, 200);
+		this.onChange = onChange;
+		
+		FlxG.gamepads.deviceConnected.add(onDeviceConnect);
+		FlxG.gamepads.deviceDisconnected.add(onDeviceDisconnect);
+		callback = onSelect;
+	}
+	
+	
+	function onDeviceConnect(gamepad:FlxGamepad):Void
```
Comment (Gama11) [inline Input/GamepadTest/source/GamepadList.hx:27]: Let's stay consistent with the API naming. (`onDeviceConnected`).

File: Input/GamepadTest/source/GamepadList.hx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	{
+		updateList();
+		select(gamepad.id);
+	}
+	
+	function onDeviceDisconnect(gamepad:FlxGamepad):Void
```
Comment (Gama11) [inline Input/GamepadTest/source/GamepadList.hx:33]: Same.

**Comment by Gama11** [comment]
There seem to be various CI failures for Haxe 3.4.7.

File: Input/GamepadTest/source/Main.hx (hunk 1 above)
Comment (Geokureli) [reply Input/GamepadTest/source/Main.hx:12]: yes

**Comment by Geokureli** [comment]
more changes incoming

File: Input/GamepadTest/source/Gamepad.hx [hunk 8]
```diff
@@ -2,195 +2,274 @@ package;
...
+		line.angle = offset.degrees;
+		line.origin.set(1, 1);
+		return line;
+	}
+	
+	inline function createStick(x:Float, y:Float, input:InputID)
+	{
+		var isLeft = input == LEFT_ANALOG_STICK;
+		var stick = createInputSprite(x, y, "Stick", input, Stick);
+		x = stick.x - this.x;// prevent double nested offset from spritegroup
```
Comment (Gama11) [inline Input/GamepadTest/source/Gamepad.hx:155]: If hacks like this are needed, should it really be a `FlxSpriteGroup`?

File: Input/GamepadTest/source/Gamepad.hx [hunk 9]
```diff
@@ -2,195 +2,274 @@ package;
...
+	function drawLine(x:Float, y:Float, offset:FlxVector, color:FlxColor):FlxSprite
+	{
+		var line = new FlxSprite(x, y);
+		line.makeGraphic(Math.floor(offset.length), 2, color);
+		line.angle = offset.degrees;
+		line.origin.set(1, 1);
+		return line;
+	}
+	
+	inline function createStick(x:Float, y:Float, input:InputID)
```
Comment (Gama11) [inline Input/GamepadTest/source/Gamepad.hx:151]: This is not a good candidate for inlining.

File: Input/GamepadTest/source/Gamepad.hx [hunk 10]
```diff
@@ -2,195 +2,274 @@ package;
...
+		createInputSprite(x - 10, y - 10, "Arrow", input, Invisible).angle = angle;
+	}
+	
+	public function setActiveGamepad(gamepad:FlxGamepad):Void
+	{
+		this.gamepad = gamepad;
+		updateLabels();
+	}
+	
+	public inline function updateLabels():Void
```
Comment (Gama11) [inline Input/GamepadTest/source/Gamepad.hx:189]: This doesn't seem like a good use of `inline` either.

File: Input/GamepadTest/source/Gamepad.hx [hunk 11]
```diff
@@ -202,72 +281,26 @@ class Gamepad extends FlxTypedGroup<FlxSprite>
...
 
-	function updateDpad()
-	{
-		var pressed = gamepad.pressed;
-
-		var left = pressed.DPAD_LEFT;
-		var right = pressed.DPAD_RIGHT;
-		var up = pressed.DPAD_UP;
-		var down = pressed.DPAD_DOWN;
+typedef Bar = { bar:FlxBar, label:FlxText };
```
Comment (Gama11) [inline Input/GamepadTest/source/Gamepad.hx:295]: Good practice to make these types private if possible.

File: Input/GamepadTest/source/Gamepad.hx (hunk 10 above)
Comment (Geokureli) [reply Input/GamepadTest/source/Gamepad.hx:189]: this one looks like a leftover from a previous implementation

File: Input/GamepadTest/source/Gamepad.hx (hunk 8 above)
Comment (Geokureli) [reply Input/GamepadTest/source/Gamepad.hx:155]: making it a FlxGroup would mean adding a different hack IMO. I will look into that option, though

File: Input/GamepadTest/source/Gamepad.hx (hunk 9 above)
Comment (Geokureli) [reply Input/GamepadTest/source/Gamepad.hx:151]: whoops! it was actually shorter to inline until I added the x = stick.x - this.x

File: Input/GamepadTest/source/Gamepad.hx (hunk 11 above)
Comment (Geokureli) [reply Input/GamepadTest/source/Gamepad.hx:295]: The subtypes? I actually didn't know that was possible until I just googled it. I wish I knew this a long time ago

**Comment by Geokureli** [comment]
resolved conversations to show that requested changes were made. not sure what the common protocol is

**Comment by Geokureli** [comment]
What target and if on html5, what browser? On [the current GamepadTest](https://haxeflixel.com/demos/GamepadTest/) on chrome, I need to interact with the gamepad to trigger the connect event and clear the "no gamepads connected" same with the new one

**Comment by Gama11** [comment]
Neko and HL.

And yeah, looks like the HTML5 behavior hasn't changed, I need to interact for the overlay to be removed both before and after these changes.

**Comment by Gama11** [comment]
Yeah, that works better.

Thanks!

**Pull request merged**